### PR TITLE
Add jit helper stub for acmp

### PR DESCRIPTION
**Pull request by tajila**
Add jit helper stub for acmp

Depends on https://github.com/eclipse/omr/pull/4841

Signed-off-by: Tobi Ajila <atobia@ca.ibm.com>

### CODE REVIEW

**Comment by tajila** [comment]
@gacholio please review these changes

File: runtime/codert_vm/cnathelp.cpp [hunk 1]
```diff
@@ -2250,6 +2250,18 @@ old_fast_jitTypeCheckArrayStoreWithNullCheck(J9VMThread *currentThread)
 	return slowPath;
 }
 
+void* J9FASTCALL
```
Comment (gacholio) [inline runtime/codert_vm/cnathelp.cpp]: I don't think this needs a return value.

File: runtime/codert_vm/cnathelp.cpp [hunk 2]
```diff
@@ -3102,6 +3114,21 @@ fast_jitTypeCheckArrayStoreWithNullCheck(J9VMThread *currentThread, j9object_t d
...
```
Comment (gacholio) [inline runtime/codert_vm/cnathelp.cpp]: If this is the direct call version, you need to return the value from the function, not use the macro.

File: runtime/codert_vm/cnathelp.cpp [hunk 3]
```diff
@@ -3357,6 +3384,8 @@ initPureCFunctionTable(J9JavaVM *vm)
 	jitConfig->old_slow_jitNewInstanceImplAccessCheck = (void*)old_slow_jitNewInstanceImplAccessCheck;
 	jitConfig->old_slow_jitTranslateNewInstanceMethod = (void*)old_slow_jitTranslateNewInstanceMethod;
 	jitConfig->old_slow_jitReportFinalFieldModified = (void*)old_slow_jitReportFinalFieldModified;
+	jitConfig->fast_jitAcmpHelper = (void*)fast_jitAcmpHelper;
```
Comment (gacholio) [inline runtime/codert_vm/cnathelp.cpp]: Does not need a table entry.

File: runtime/jilgen/jilconsts.c [hunk 4]
```diff
@@ -632,6 +632,8 @@ writeConstants(OMRPortLibrary *OMRPORTLIB, IDATA fd)
 			writeConstant(OMRPORTLIB, fd, "J9TR_JitConfig_old_slow_jitReportStaticFieldRead", offsetof(J9JITConfig, old_slow_jitReportStaticFieldRead)) |
 			writeConstant(OMRPORTLIB, fd, "J9TR_JitConfig_old_slow_jitReportStaticFieldWrite", offsetof(J9JITConfig, old_slow_jitReportStaticFieldWrite)) |
 
+			writeConstant(OMRPORTLIB, fd, "J9TR_JitConfig_fast_jitAcmpHelper", offsetof(J9JITConfig, fast_jitAcmpHelper)) |
```
Comment (gacholio) [inline runtime/jilgen/jilconsts.c]: Not needed.

File: runtime/oti/j9nonbuilder.h [hunk 5]
```diff
@@ -3504,6 +3504,8 @@ typedef struct J9JITConfig {
 	void *old_slow_jitNewInstanceImplAccessCheck;
 	void *old_slow_jitTranslateNewInstanceMethod;
 	void *old_slow_jitReportFinalFieldModified;
+	void *fast_jitAcmpHelper;
```
Comment (gacholio) [inline runtime/oti/j9nonbuilder.h]: Not needed.

File: runtime/vm/valueTypeHelpers.cpp [hunk 6]
```diff
@@ -54,4 +54,29 @@ defaultValueWithUnflattenedFlattenables(J9VMThread *currentThread, J9Class *claz
...
 }
+
+BOOLEAN
+valueTypeCapableAcmp(J9VMThread *currentThread, j9object_t lhs, j9object_t rhs)
+{
+#if defined(J9VM_OPT_VALHALLA_VALUE_TYPES)
+		bool acmpResult = false;
+		if (rhs == lhs) {
+			acmpResult = true;
+		} else if ((NULL == rhs) || (NULL == lhs)) {
```
Comment (gacholio) [inline runtime/vm/valueTypeHelpers.cpp]: Maybe write this as an else followed by the if, which wraps the class fetch/call.

**Comment by tajila** [comment]
@gacholio Ive addressed your comments

**Comment by gacholio** [comment]
Compile failing in the JIT.

**Comment by tajila** [comment]
> Compile failing in the JIT.

it needs to be compiled with the dependant omr changes in https://github.com/eclipse/omr/pull/4841/

**Comment by tajila** [comment]
you can do something like `Jenkins test sanity x86-64_linux_xl,s390x_linux,ppc64le_linux_xl jdk8 depends eclipse/omr#4841`

**Comment by gacholio** [comment]
Jenkins test sanity x86-64_linux_xl,s390x_linux,ppc64le_linux_xl jdk8 depends eclipse/omr#4841

**Comment by gacholio** [comment]
Jenkins test sanity all jdk8 depends eclipse/omr#4841

**Comment by gacholio** [comment]
The test failures look like infra issues. Let me know when the OMR change is in and I'll merge.

**Comment by tajila** [comment]
@gacholio the OMR changes have been merged

**Comment by gacholio** [comment]
Has it been through OMR acceptance?

**Comment by pshipton** [comment]
> Has it been through OMR acceptance?

The build is running now, should be promoted in about another hour.

**Comment by pshipton** [comment]
> Has it been through OMR acceptance?

Promoted now.

**Pull request merged**